jplag.frontend-utils/src/main/java/de/jplag/TokenPrinter.java

```diff
@@ -104,10 +106,15 @@ public static String printTokens(TokenList tokens, Collection<File> files, File
             }
 
             // New code line:
-            if (token.getLine() > lineIndex) {
-                lineIndex = token.getLine();
+            boolean previousLineEmpty = false;
+            while (token.getLine() > lineIndex) {
+                lineIndex++;
                 columnIndex = 1;
 
+                boolean currentLineEmpty = token.getLine() > lineIndex;
+                if (currentLineEmpty && !PRINT_EMPTY_LINES)
+                    continue;
+
                 String fileName = token.getFile().isEmpty() ? root.getName() : token.getFile();
                 fileName = fileName + suffix.orElse("");
                 if (!filesToLines.containsKey(fileName)) {
```

Comment:
@jplag/maintainer 
I'm having a go at a refactoring, but I'm having trouble understanding why using the root folder as a replacement file name would be valid. Can someone explain?

Reply:
Don't know neither but if you refactor it anyways, I would opt for doing it properly, i.e.
```Java
for (File file: files) {
  List<String> lines = getLinesForFile(file); //invented method but probably better than loading all lines upfront
  builder.append(System.lineSeparator());
  builder.append(file.getName());
  for (String line: lines) {
    / print lines and tokens
  }
}
```
and then you don't need the fallback anymore.

Reply:
Whenever your submission folder contains single files instead of folders, the root is the name of the file, and `Token.getFile()` is `""`.

This is a valid folder structure:
```
input/
|- submissionA.java
|- submissionB.java
|- submissionC.java
```